From d143af396f41d18186abffda3db6ff854dae253f Mon Sep 17 00:00:00 2001
From: Hugues Fruchet <hugues.fruchet@foss.st.com>
Date: Wed, 10 Dec 2025 14:36:44 +0100
Subject: [PATCH] media: verisilicon: g1: jpeg: abort on timeout

G1 hardware timeout feature (see G1_REG_CONFIG_DEC_TIMEOUT_E)
resets hardware without checking the presence of pending bus
transactions: this can hang the bus in an unrecoverable way.

As a workaround, disable hardware timeout feature and
replace it by a software timeout.
This is implemented by checking if G1 is idle before each
run, if it is not the case hardware is aborted: this will
clean all the bus left transactions.

Change-Id: I391341ad5f5ae9c11bb9177515aa60686dc09a67
Signed-off-by: Hugues Fruchet <hugues.fruchet@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/519341
Reviewed-by: Hugues FRUCHET <hugues.fruchet@st.com>
Reviewed-by: Philippe CORNU <philippe.cornu@foss.st.com>
Domain-Review: Philippe CORNU <philippe.cornu@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Tested-by: Hugues FRUCHET <hugues.fruchet@st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
---
 drivers/media/platform/verisilicon/hantro_g1_jpeg_dec.c | 2 ++
 1 file changed, 2 insertions(+)

--- a/drivers/media/platform/verisilicon/hantro_g1_jpeg_dec.c
+++ b/drivers/media/platform/verisilicon/hantro_g1_jpeg_dec.c
@@ -239,6 +239,8 @@ int hantro_g1_jpeg_dec_run(struct hantro
 	int ret;
 	u32 i;
 
+	hantro_g1_check_idle(vpu);
+
 	hantro_start_prepare_run(ctx);
 
 	src_buf = hantro_get_src_buf(ctx);
